quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// s2n-quic's default implementation of the datagram component
+
+use crate::{
+    datagram::unreliable_datagram::{ConnectionInfo, Endpoint, Packet, Receiver, Sender},
+    time::Timestamp,
+};
+use alloc::collections::VecDeque;
+use bytes::Bytes;
+use core::time::Duration;
+
+#[derive(Debug, Default)]
+pub struct Disabled;
+
+// Default capacity is zero right now since the feature is currently disabled.
+const DEFAULT_CAPACITY: usize = 0;
+
+impl Endpoint for Disabled {
+    type Sender = DefaultSender;
+    type Receiver = DisabledReceiver;
+
+    fn create_connection(&mut self, _info: &ConnectionInfo) -> (Self::Sender, Self::Receiver) {
+        let queue = VecDeque::with_capacity(DEFAULT_CAPACITY);
+        (
+            DefaultSender {
+                queue,
+                prioritize_datagrams: true,
+            },
+            DisabledReceiver,
+        )
+    }
+}
+
+pub struct DefaultSender {
+    queue: VecDeque<Datagram>,
+    prioritize_datagrams: bool,
+}
+
+struct Datagram {
+    data: Bytes,
+    creation_time: Timestamp,
+}
+
+// Expiration in ms of datagrams on the queue. If datagrams are older than 3 ms,
+// consider them expired and don't send them.
+const DATAGRAM_EXPIRATION: u64 = 3;
+
+impl Sender for DefaultSender {
+    fn on_transmit<P: Packet>(&mut self, packet: &mut P) {
+        // Alternate between sending datagrams and ceding that space to pending
+        // stream data
+        let cede_space = packet.has_pending_streams() && !self.prioritize_datagrams;
```

Comment:
Is this going to cause sync issues with the control frame logic? As in is there a possibility that when control frames prioritize datagrams, then this prioritizes streams? And visa versa.

Reply:
Yes currently that situation is a possibility. I could add a function that queries whether or not we're sending datagrams first on this packet, so the application can sync with that if it wants to. Is that what you're thinking of?

Reply:
Yeah if packet had a `prioritize_datagrams` function the sender component wouldn't have to duplicate the logic.